---
title: "Tutorial 4"
author: "Feras Machour"
date: "11/18/2021"
output: github_document
editor_options: 
  chunk_output_type: console
---
[GitHub](https://github.com/ferasmachour/Genomic_Data_Science_Course.git)
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load packages, include=FALSE}
library(compGenomRData)
library(DESeq2)
library(gProfileR)
library(gage)
library(topGO)
library(pheatmap)
library(org.Hs.eg.db)
library(biomaRt)
library(msigdbr)


```

```{r load data and perform DESeq2 analysis}
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
package = "compGenomRData")
counts <- as.matrix(read.table(counts_file, header = T, sep = '\t'))
#remove the 'width' column
countData <- as.matrix(subset(counts, select = c(-width)))
#define the experimental setup
colData <- read.table(coldata_file, header = T, sep = '\t',
stringsAsFactors = TRUE)
#define the design formula
designFormula <- "~ group"
#create a DESeq dataset object from the count matrix and the colData
dds <- DESeqDataSetFromMatrix(countData = countData,
colData = colData,
design = as.formula(designFormula))
dds <- DESeq(dds)
DEresults = results(dds, contrast = c("group", 'CASE', 'CTRL'))

```

```{r prepare data for gene-set enrichment analysis}
#remove genes with NA values
DEresults_clean <- DEresults[!is.na(DEresults$padj),]

#select genes with adjusted p-values below 0.1
DE <- DEresults_clean[DEresults_clean$padj < 0.1,]
#select genes with absolute log2 fold change above 1 (two-fold change)
DE <- DE[abs(DE$log2FoldChange) > 1,]
#get the list of genes of interest
genesOfInterest <- rownames(DE)
```

#####  gProfileR
1. Re-run gProfileR, this time using pathway annotations such as KEGG, REACTOME, and protein complex databases such as CORUM, in addition to the GO terms. Sort the resulting tables by columns precision and/or recall. How do the top GO terms change when sorted for precision, recall, or p.value? hint: use order() for sorting


```{r gProfileR with GO enrichment}
goResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'GO',
hier_filtering = 'moderate')
head(goResults[,1:13])

# order by precision: we get the most true positives from the selected elements first
head(goResults[order(goResults$precision,decreasing = TRUE),1:13])

# order by recall: we get the most true positives from the relevant elements first
head(goResults[order(goResults$recall,decreasing = TRUE),1:13])
# order by p.value: get the most significant first
head(goResults[order(goResults$p.value,decreasing = FALSE),1:13])


```





```{r gProfileR with KEGG}
KEGGResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'KEGG',
hier_filtering = 'moderate')
head(KEGGResults[order(KEGGResults$p.value,decreasing = FALSE),1:13])
```


```{r gProfileR with REACTOME}
REACTOMEResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'REAC',
hier_filtering = 'moderate')
head(REACTOMEResults[order(REACTOMEResults$p.value,decreasing = FALSE),1:13])
```

```{r gProfileR with CORUM}
CORUMResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'CORUM',
hier_filtering = 'moderate')
# no significant enrichment found with the CORUM database
head(CORUMResults[order(CORUMResults$p.value,decreasing = FALSE),1:13])
```



#### Gene set enrichment analysis

We use the bioconductor package gage to demonstrate how to do GSEA using normalized expression data of the samples as input.

```{r}
#Let's define the first gene set as the list of genes from one of the
#significant GO terms found in the GO analysis. order go results by pvalue
goResults <- goResults[order(goResults$p.value),]

#restrict the terms that have at most 100 genes overlapping with the query
go <- goResults[goResults$overlap.size < 100,]

# use the top term from this table to create a gene set
geneSet1 <- unlist(strsplit(go[1,]$intersection, ','))


#table get normalized counts from DESeq2 results
normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)

#Define another gene set by just randomly selecting 25 genes from the counts
geneSet2 <- sample(rownames(normalizedCounts), 25)

# geneSets list

geneSets <- list('top_GO_term' = geneSet1,
                 'random_set' = geneSet2)

# Using the defined gene sets, weâ€™d like to do a group comparison between the case
# samples with respect to the control samples.

#Use the normalized counts to carry out a GSEA.
gseaResults <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'as.group')
# Up-regulation statistics
gseaResults$greater 
# Down-regulation statistics
gseaResults$less

# There is significant up-regulation of the most enriched GOTerm in the CASE vs CONTROL


# Visualize the expression of these genes by heatmap
# get the expression data for the gene set of interest
M <- normalizedCounts[rownames(normalizedCounts) %in% geneSet1, ]
# log transform the counts for visualization scaling by row helps visualizing
# relative change of expression of a gene in multiple conditions
pheatmap(log2(M+1),
annotation_col = colData,
show_rownames = TRUE,
fontsize_row = 8,
scale = 'row',
cutree_cols = 2,
cutree_rows = 2)
# Looking at the heatmap we can see that most genes in the most enriched GOterm are indeed upregulated 
```

2. Repeat the gene set enrichment analysis by trying different options for the `compare` argument of the `GAGE:gage` function. How do the results differ? [Difficulty: **Beginner**]

```{r gage with different compare options}
# compare = 'paired' - every case sample is paired with a control sample - one p-value for each pair comparison
gseaResults.paired <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'paired')
gseaResults.paired

# compare = 'unpaired' - one-on-one comparison between all possible CONTROL and CASE combinations
gseaResults.unpaired <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'unpaired')
gseaResults.unpaired

# compare = '1ongroup', comparison between one CASE column at a time vs the average of all CONTROL columns.

gseaResults.1ongroup <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = '1ongroup')
gseaResults.1ongroup

# we get sligthly different p-values for each compare method but the trend is the same

```


3. Make a scatter plot of GO term sizes and obtained p-values by setting the `gProfiler::gprofiler` argument `significant = FALSE`. Is there a correlation of term sizes and p-values? (Hint: Take -log10 of p-values). If so, how can this bias be mitigated? [Difficulty: **Intermediate**]


```{r correlation term size vs p-value}
# redo the GO analysis while including all terms (also non significant)
goResults_all <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'GO',
hier_filtering = 'moderate', significant = FALSE)
head(goResults_all[,1:13])


plot(goResults_all$term.size,-log10(goResults_all$p.value),xlab = "term size", ylab = "-log10(p.value)", main = "Correlation between term size and p-value", type = "p")
# The bigger the term size, the more likely it will be significant
# To mitigate this, we can look at alternative metrics such as fold-enrichment and false-discovery

```

4. Do a gene-set enrichment analysis using gene sets from top 10 GO terms. [Difficulty: **Intermediate**]

```{r GSEA analysis top 10 GO terms}
Top10geneSet <- strsplit(go[1:10,]$intersection, ',')
Top10geneSet <- setNames(Top10geneSet, go[1:10,]$term.name)
head(Top10geneSet)



gseaResults <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = Top10geneSet, compare = 'as.group')
# Up-regulation statistics
gseaResults$greater 
# Down-regulation statistics
gseaResults$less

# There is significant up-regulation in the top 10 GOTerms in the CASE vs CONTROL



expression <- normalizedCounts[rownames(normalizedCounts) %in% unlist(Top10geneSet,use.names=FALSE),]

pheatmap(log2(expression+1),
annotation_col = colData,
show_rownames = FALSE,
fontsize_row = 8,
scale = 'row',
cutree_cols = 2,
cutree_rows = 2)
# heatmap reveals that not all genes from these GOterms are upregulated (the majority are upregulated)
```

##### 5. What are the other available R packages that can carry out gene set enrichment analysis for RNA-seq datasets? [Difficulty: **Intermediate**]
##### ClusterProfiler, fgsea, goseq ...


#####6.  Use the topGO package (https://bioconductor.org/packages/release/bioc/html/topGO.html) to re-do the GO term analysis. Compare and contrast the results with what has been obtained using the `gProfileR` package. Which tool is faster, `gProfileR` or topGO? Why? [Difficulty: **Advanced**]

```{r topGO}
geneNames <- rownames(DEresults_clean)
geneList <- factor(as.integer(geneNames %in% genesOfInterest))
names(geneList) <- geneNames
str(geneList)

 topGOdata <- new("topGOdata", description = "Tutorial 4", ontology = "BP", allGenes = geneList, annot = annFUN.org, ID = "symbol", mapping = "org.Hs.eg.db",nodeSize=10)

 topGOdata
 
 resultFisher <- runTest(topGOdata, algorithm = "classic", statistic = "fisher")
 resultFisher
 resultKS <- runTest(topGOdata, algorithm = "classic", statistic = "ks")
resultKS.elim <- runTest(topGOdata, algorithm = "elim", statistic = "ks")
 
# Look at the top 10 most significant GO terms for the topGO analysis

 allRes <- GenTable(topGOdata, classicFisher = resultFisher, classicKS = resultKS, elimKS = resultKS.elim, orderBy = "elimKS", ranksOf = "classicFisher", topNodes = 10)
 allRes
 
 # rerun gprofiler on GO:BP with the same minimum term size to compare with topGO
 goResults_BP <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'GO:BP',
hier_filtering = 'moderate',min_set_size = 10)
 goResults_BP <- goResults_BP[order(goResults_BP$p.value),]
  goResults_BP[1:10,1:13]
  
  
# gPofiler is much faster than topGO due to the difference in statistical tests
# The results are significantly different between the two methods
```

##### 7. Given a gene set annotated for human, how can it be utilized to work on _C. elegans_ data? (Hint: See `biomaRt::getLDS`). [Difficulty: **Advanced**]
```{r getLDS}
httr::set_config(httr::config(ssl_verifypeer = FALSE))
mart = useMart('ensembl')

human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
c.elegans = useMart("ensembl", dataset = "celegans_gene_ensembl")
LDS <- getLDS(attributes = c("hgnc_symbol","chromosome_name", "start_position"), filters = "hgnc_symbol", values = genesOfInterest, mart = human, attributesL = c("chromosome_name","start_position","external_gene_name","description"), martL = c.elegans)
head(LDS)


```

#####  8. Import curated pathway gene sets with Entrez identifiers from the [MSIGDB database](http://software.broadinstitute.org/gsea/msigdb/collections.jsp) and re-do the GSEA for all curated gene sets. [Difficulty: **Advanced**]

```{r curated pathways}
pathway_genesets = msigdbr(species = "Homo sapiens", category = "C2" , subcategory = "CP")
head(pathway_genesets)

geneSets_names <- unique(pathway_genesets$gs_name)



pathway_genesets_names <- sapply(unique(pathway_genesets$gs_name), function(x) list(pathway_genesets$gene_symbol[pathway_genesets$gs_name == x]))

gseaResults_pathways <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = pathway_genesets_names, compare = 'as.group')
head(gseaResults_pathways)

expression_pathways <- normalizedCounts[rownames(normalizedCounts) %in% unlist(pathway_genesets_names,use.names=FALSE),]
expression_pathways = expression_pathways[rowSums(expression_pathways)!=0,]
pheatmap(log2(expression_pathways+1),
annotation_col = colData,
show_rownames = FALSE,
fontsize_row = 8,
scale = 'row',
cutree_cols = 2,
cutree_rows = 2, main = "Curated Pathways")

```

[link to github repository](https://github.com/ferasmachour/Genomic_Data_Science_Course.git)