---
title: "Tutorial 3"
author: "Feras Machour"
date: "11/12/2021"
output: github_document
editor_options: 
  chunk_output_type: console
---
[GitHub](https://github.com/ferasmachour/Genomic_Data_Science_Course.git)
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##### Import libraries and dataset
```{r import libs and data , include=FALSE}
library(compGenomRData)
library(DESeq2)
library(stats)
library(edgeR)
library(compcodeR)
library(ggplot2)

counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
package = "compGenomRData")


```

##### 1. Perform differential expression analysis using DESeq2

```{r DESeq2 analysis}

counts <- read.table(counts_file, header = T, sep = '\t')
#remove the 'width' column
countData <- as.matrix(subset(counts, select = c(-width)))

colData <- read.table(coldata_file, header = T, sep = '\t',
                      stringsAsFactors = TRUE) 



#create a DESeq dataset object from the count matrix and the colData
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = ~ group)

#print dds object to see the contents
print(dds)
nrow(dds)
#Filter out genes with low counts.
filteration <- rowSums(DESeq2::counts(dds)) >= 100

dds <- dds[filteration, ]
nrow(dds)
dds <- DESeq(dds)

#Run DESeq2 contrasting the CASE sample with CONTROL samples
DEresults = results(dds, contrast = c("group", 'CASE', 'CTRL'))
DEresultsDf <-as.data.frame(DEresults)


ggplot()+
  geom_point(data = DEresultsDf, 
             mapping = aes(x = log2FoldChange, y = -log10(pvalue)), 
             col = "grey80", size =1) +
  
  geom_point(data = subset(DEresultsDf, log2FoldChange > 1 & pvalue < 0.05), 
             aes(log2FoldChange, -log10(pvalue)), color = "red", size =1)+
  geom_point(data = subset(DEresultsDf, log2FoldChange < -1 & pvalue < 0.05), 
            aes(log2FoldChange, -log10(pvalue)), color = "steelblue2", size =1)+
  theme_bw()+
  theme(legend.title =element_blank())+
  labs(x = "log2 (Fold change of case vs. control)", y= "-log10(P-value)")


# The up-regulated genes in case samples are in red
#The down-regulated genes in control samples are in blue


```

##### 2. Use DESeq2::plotDispEsts to make a dispersion plot

```{r Dispersion}

plotDispEsts(dds)

```

##### 3. lfcThreshold arguement

```{r lfcThreshold arguement}
# The default value of lfcThreshold is 0

# lfcThreshold specifies the log2(foldchange) value of the null hypothesis



# changing it to 1 would test the statistical significance of the foldchange being different than 1 rather than 0


sig <- subset(DEresultsDf, padj < 0.05)
# number of significant DE genes with lfcThreshold = 0 (default)

nrow(sig) 

DEresults_lfcThreshold = results(dds, contrast = c("group", 'CASE', 'CTRL'), lfcThreshold = 1)

DEresults_lfcThresholdDf <-as.data.frame(DEresults_lfcThreshold)
sig_lfcThresholdDf <- subset(DEresults_lfcThresholdDf, padj < 0.05)
# number of significant DE genes with lfcThreshold = 1
nrow(sig_lfcThresholdDf)
```

##### 4. Independent filtering
###### without independent filtering, many discoveries will be lost due to multiple testings

##### 5. Re-do the differential expression analysis using the edgeR package 

```{r edgeR analysis}
library(statmod)
group=factor(c("CASE", "CASE", "CASE", "CASE", "CASE", "CONTROL","CONTROL","CONTROL","CONTROL","CONTROL"))
DGEList <- DGEList(counts=as.matrix(countData), group=group)
print(DGEList$samples)

# filtering and normalization
dim(DGEList$counts)

cpm <- cpm(DGEList)
head(cpm)
apply(DGEList$counts, 2, sum)
# since the libraries have a large number of reads, we chose a small cpm threshold that matches the number of filtered genes from the DESeq2 count data 
# alternatively, this filtering could be done similar to the way it was done for DESeq2 (wanted to try the cpm approach)

keep <- rowSums(cpm(DGEList)>0.01) >= 5
DGEList <- DGEList[keep,]
dim(DGEList$counts)
# Recompute the library sizes
DGEList$samples$lib.size <- colSums(DGEList$counts)
print(DGEList$samples)

# Compute normalization factor
DGEList <- calcNormFactors(DGEList)
print(DGEList$samples)

plotMDS(DGEList, pch=20, col=as.numeric(DGEList$samples$group))
legend("bottomleft", as.character(unique(DGEList$samples$group)), col=1:3, pch=20)



# compute dispersion

DGEList <- estimateDisp(DGEList,verbose=T)
plotBCV(DGEList)

# Differential expression analysis

fit <- glmFit(DGEList)

# number of significant DE genes with edgeR


DE <- exactTest(DGEList, pair=c(1,2))
topTags(DE)
summary(decideTests(DE))
DE <- topTags(DE, n = "Inf")

sig_edgeR <- subset(DE$table, FDR <= 0.05)
print(nrow(sig_edgeR))
# number of significant DE genes with DESeq2

sig_DESeq <- subset(DEresultsDf, padj <= 0.05)
print(nrow(sig_DESeq))


# similar number of DE genes

library(VennDiagram)

venn <- venn.diagram(
  x = list(row.names(sig_edgeR), row.names(sig_DESeq)),
  category.names = c("edgeR" , "DESeq2 " ),
  filename = NULL,
  lwd = 2,
  lty = 'blank',
  fill = c("#440154ff" , '#21908dff')
  )
grid.newpage()
grid.draw(venn)


```

##### 6. Use the compcodeR package to run the differential expression analysis using at least three different tools

```{r compcodeR}
info.parameters <- list(dataset = "countdata", uID = "123456")
colnames(colData)[2] <- "condition"

cpd <- compData(count.matrix = countData, 
                sample.annotations = colData, 
                info.parameters = info.parameters)
saveRDS(cpd, file = "cpd.rds")

check_compData(cpd)

runDiffExp("cpd.rds", 
           result.extent = "voom.limma", Rmdfunction = "voom.limma.createRmd", 
           output.directory = ".", norm.method = "TMM")
runDiffExp("cpd.rds", 
           result.extent = "DESeq2", Rmdfunction = "DESeq2.createRmd", 
           output.directory = ".", fit.type = "parametric",test = "Wald")
runDiffExp("cpd.rds", 
           result.extent = "ttest", Rmdfunction = "ttest.createRmd", 
           output.directory = ".", norm.method = "TMM")
voom.limma = readRDS("cpd_voom.limma.rds")
DESeq2 = readRDS("cpd_DESeq2.rds")
ttest = readRDS("cpd_ttest.rds")

voom.limma <- as.data.frame(voom.limma@result.table)
sig_limma = subset(voom.limma, adjpvalue <= 0.05)
nrow(sig_limma)

DESeq2 <- as.data.frame(DESeq2@result.table)
sig_DESeq2 = subset(DESeq2, adjpvalue <= 0.05)
nrow(sig_DESeq2)

ttest <- as.data.frame(ttest@result.table)
sig_ttest = subset(ttest, adjpvalue <= 0.05)
nrow(sig_ttest)


library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")
venn <- venn.diagram(
  x = list(row.names(sig_limma), row.names(sig_DESeq2), row.names(sig_ttest)),
  category.names = c("limma" , "DESeq2 " , " ttest"),
  filename = NULL,
  lwd = 2,
  lty = 'blank',
  fill = myCol
  )
grid.newpage()
grid.draw(venn)

```
[link to github repository](https://github.com/ferasmachour/Genomic_Data_Science_Course.git)