---
title: "Tutorial 2"
author: "Feras Machour"
date: "11/4/2021"
output: html_document
editor_options: 
  chunk_output_type: inline
---
[GitHub](https://github.com/ferasmachour/Genomic_Data_Science_Course.git)
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###### Import counts data and design

```{r import count and design data}
library(compGenomRData)

counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
package = "compGenomRData") 
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
package = "compGenomRData") # get the filepath

countdata = read.table(counts_file,header=TRUE, sep = '\t') # import data into R
countdata_mat = as.matrix(countdata) # convert dataframe to matrix
head(countdata_mat)

coldata = read.table(coldata_file,header=TRUE,sep='\t')
head(coldata)

```

#### 1. Normalize the counts using the TPM approach

```{r TPM calculation}
geneLengths = as.vector(countdata_mat[,"width"])

readcounts = subset(countdata_mat, select = c(-width))

RPK = apply(readcounts, 2 , function(x){x/(geneLengths/1000)}) # Calculate RPK

per_million = colSums(RPK)/10^6 # Calculate per million scaling factor

TPM = mapply('/', data.frame(RPK) ,per_million) 
row.names(TPM) <- row.names(RPK) # Calculate TPM

head(TPM)

colSums(TPM) # Check whether columns are correctly normalized


```

#### 2. Plot a heatmap of the top 500 most variable genes. Compare with the heatmap obtained using the top 100 most variable genes. 

```{r 2}

variance = apply(TPM , 1 , var)
variance = variance[order(variance, decreasing = 1)] # sort variance
Top500 <- names(variance[1:500])
Top100 <- names(variance[1:100])
library(pheatmap)
library(gridExtra)

pheatmap(TPM[Top500,] , show_rownames = FALSE , scale = "row", main = "Top 500" , annotation_col = coldata)
pheatmap(TPM[Top100,] , show_rownames = FALSE , scale = "row", main = "Top 100" , annotation_col = coldata)

```

#### 3. Re-do the heatmaps setting the scale argument to none, and column. Compare the results with scale = 'row'.

```{r 3}

pheatmap(TPM[Top500,] , show_rownames = FALSE , scale = "row", main = "Top 500 - scale by rows" , annotation_col = coldata["group"])
pheatmap(TPM[Top500,] , show_rownames = FALSE , scale = "column", main = "Top 500 - scale by columns" , annotation_col = coldata["group"])
pheatmap(TPM[Top500,] , show_rownames = FALSE , scale = "none", main = "Top 500 - no scaling" , annotation_col = coldata["group"])

```

#### 4. Draw a correlation plot for the samples depicting the sample differences as ‘ellipses’, drawing only the upper end of the matrix, and order samples by hierarchical clustering results based on average linkage clustering method.

```{r 4}

library(stats)
library(corrplot)

correlationMatrix <- cor(TPM)
head(correlationMatrix)


corrplot(correlationMatrix, method = "ellipse" , hclust.method = "average" , order = "hclust", addrect = 2, addCoef.col = 'green', number.cex = 0.8,tl.cex = 0.5, tl.col = "black", type = "upper")
pheatmap(correlationMatrix, annotation_col = coldata , cutree_cols = 2)


```

#### 5. How else could the count matrix be subsetted to obtain quick and accurate clusters? Try selecting the top 100 genes that have the highest total expression in all samples and re-draw the cluster heatmaps and PCA plots.

```{r 5}
library(ggplot2)
library(stats)
library(pheatmap)
library(ggfortify)
library(corrplot)

gene_average_exp = apply(TPM , 1 , mean)
gene_average_exp = gene_average_exp[order(gene_average_exp,decreasing=1)]

High_exp = names(gene_average_exp[1:100])

pheatmap(TPM[High_exp,] , show_rownames = FALSE , scale = "row", main = "100 highest expression genes" , annotation_col = coldata["group"])

# Transform data for PCA
transpose_TPM = t(TPM[High_exp,])
transpose_TPM = log2(transpose_TPM + 1) 
pcaResults = prcomp(transpose_TPM)

autoplot(pcaResults, data = coldata, colour = 'group', main = "PCA - highest expression genes")

# CorrelationMatrix for the highest 100 genes

correlationMatrix_High <- cor(TPM[High_exp,])

pheatmap(correlationMatrix_High, annotation_col = coldata , cutree_cols = 2)
corrplot(correlationMatrix_High, method = "ellipse" , hclust.method = "average" , order = "hclust", addrect = 2, addCoef.col = 'green', number.cex = 0.8,tl.cex = 0.5, tl.col = "black", type = "upper", add = FALSE)

```

#### 6. Add an additional column to the annotation data.frame object to annotate the samples and use the updated annotation data.frame to plot the heatmaps

```{r 6}
# assign batches based on correlation data
batch = c("1","3","1","3","2","1","3","1","3","2")
coldata = cbind(coldata,batch)
pheatmap(correlationMatrix, annotation_col = coldata , cutree_cols = 2)


transpose_TPM = t(TPM)
transpose_TPM = log2(transpose_TPM + 1) 
pcaResults = prcomp(transpose_TPM)
autoplot(pcaResults, data = coldata, colour = 'batch', shape = 'group' , main = "PCA")


```

#### 7. Try making the heatmaps using all the genes in the count table, rather than sub-selecting.
```{r 7}
library(tidyr)

# pheatmap(TPM , show_rownames = FALSE , main = "Heatmap of all the genes" , annotation_col = coldata , scale = "row") # doesn't work




is.na(TPM) %>% table() 
dim(TPM)

# no missing values

length(variance[variance==0])
# list contains genes with 0 variance - probably those with 0 reads across all samples
# remove genes with 0 reads across all samples
TPM_clean = TPM[rowSums(TPM)!=0,]
pheatmap(TPM_clean , show_rownames = FALSE , main = "Heatmap of all the genes" , annotation_col = coldata , scale = "row")
```

[link to github repository](https://github.com/ferasmachour/Genomic_Data_Science_Course.git)